mg/agent.go

```diff
@@ -214,21 +222,17 @@ func (ag *Agent) shutdown() {
 	}
 	sd.closed = true
 
-	// shutdown sequence:
-	// * stop incoming requests
-	// * wait for all reqs to complete
-	// * tell reducers we're shutting down
-	// * stop outgoing responses
-	// * tell the world we're done
-
 	// defers because we want *some* guarantee that all these steps will be taken
 	defer close(sd.done)
 	defer ag.stdout.Close()
+	defer ag.stderr.Close()
 	defer ag.Store.dispatch(Shutdown{})
 	defer ag.wg.Wait()
 	defer ag.stdin.Close()
 }
 
+// NewAgent returns an error if an invalid codec has been passed. See
+// AgentConfig.Codec for codecs.
```

Comment:
Do you reckon the document should mention `the object might not be fully usable, but it will not be nil` or something like that? Otherwise it is not clear to others.

Reply:
Thinking about it some more, I think it's worth just switching the codec to `json` and documenting that it will use json, that way the object is in completely valid state.

I'd say something like

    NewAgent returns a new Agent, initialised using the settings in cfg.
    If cfg.Codec is invalid (see CodecNames), `DefaultCodec` will be used as the codec and an error returned.
    An initialised, usable agent object is always returned.

Reply:
Would you like me to apply the new changes based on the default codec here in this PR?